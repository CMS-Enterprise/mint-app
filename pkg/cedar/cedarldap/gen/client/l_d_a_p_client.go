// Code generated by go-swagger; DO NOT EDIT.

package client

// This file was generated by the swagger tool.
// Editing this file might prove futile when you re-run the swagger generate command

import (
	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"

	strfmt "github.com/go-openapi/strfmt"

	"github.com/cmsgov/easi-app/pkg/cedar/cedarldap/gen/client/operations"
)

// Default l d a p HTTP client.
var Default = NewHTTPClient(nil)

const (
	// DefaultHost is the default Host
	// found in Meta (info) section of spec file
	DefaultHost string = "webmethods-apigw.cedarimpl.cms.gov"
	// DefaultBasePath is the default BasePath
	// found in Meta (info) section of spec file
	DefaultBasePath string = "/gateway/LDAP/1.0"
)

// DefaultSchemes are the default schemes found in Meta (info) section of spec file
var DefaultSchemes = []string{"http"}

// NewHTTPClient creates a new l d a p HTTP client.
func NewHTTPClient(formats strfmt.Registry) *LDAP {
	return NewHTTPClientWithConfig(formats, nil)
}

// NewHTTPClientWithConfig creates a new l d a p HTTP client,
// using a customizable transport config.
func NewHTTPClientWithConfig(formats strfmt.Registry, cfg *TransportConfig) *LDAP {
	// ensure nullable parameters have default
	if cfg == nil {
		cfg = DefaultTransportConfig()
	}

	// create transport and client
	transport := httptransport.New(cfg.Host, cfg.BasePath, cfg.Schemes)
	return New(transport, formats)
}

// New creates a new l d a p client
func New(transport runtime.ClientTransport, formats strfmt.Registry) *LDAP {
	// ensure nullable parameters have default
	if formats == nil {
		formats = strfmt.Default
	}

	cli := new(LDAP)
	cli.Transport = transport

	cli.Operations = operations.New(transport, formats)

	return cli
}

// DefaultTransportConfig creates a TransportConfig with the
// default settings taken from the meta section of the spec file.
func DefaultTransportConfig() *TransportConfig {
	return &TransportConfig{
		Host:     DefaultHost,
		BasePath: DefaultBasePath,
		Schemes:  DefaultSchemes,
	}
}

// TransportConfig contains the transport related info,
// found in the meta section of the spec file.
type TransportConfig struct {
	Host     string
	BasePath string
	Schemes  []string
}

// WithHost overrides the default host,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithHost(host string) *TransportConfig {
	cfg.Host = host
	return cfg
}

// WithBasePath overrides the default basePath,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithBasePath(basePath string) *TransportConfig {
	cfg.BasePath = basePath
	return cfg
}

// WithSchemes overrides the default schemes,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithSchemes(schemes []string) *TransportConfig {
	cfg.Schemes = schemes
	return cfg
}

// LDAP is a client for l d a p
type LDAP struct {
	Operations *operations.Client

	Transport runtime.ClientTransport
}

// SetTransport changes the transport on the client and all its subresources
func (c *LDAP) SetTransport(transport runtime.ClientTransport) {
	c.Transport = transport

	c.Operations.SetTransport(transport)

}
