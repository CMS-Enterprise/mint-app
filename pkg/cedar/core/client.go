package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
