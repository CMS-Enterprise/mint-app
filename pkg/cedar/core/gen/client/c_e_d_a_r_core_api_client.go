// Code generated by go-swagger; DO NOT EDIT.

package client

// This file was generated by the swagger tool.
// Editing this file might prove futile when you re-run the swagger generate command

import (
	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"

	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/authority_to_operate"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/budget"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/component"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/contract"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/cost_type"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/data_center"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/deployment"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/domain_model"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/enumeration"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/exchange"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/person"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/role"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/stakeholder"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/support_contact"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/system"
	"github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/user"
)

// Default c e d a r core API HTTP client.
var Default = NewHTTPClient(nil)

const (
	// DefaultHost is the default Host
	// found in Meta (info) section of spec file
	DefaultHost string = "cdrprdapigw1.awscloud.cms.local:5555"
	// DefaultBasePath is the default BasePath
	// found in Meta (info) section of spec file
	DefaultBasePath string = "/gateway/CEDAR Core API/1.0.0"
)

// DefaultSchemes are the default schemes found in Meta (info) section of spec file
var DefaultSchemes = []string{"https"}

// NewHTTPClient creates a new c e d a r core API HTTP client.
func NewHTTPClient(formats strfmt.Registry) *CEDARCoreAPI {
	return NewHTTPClientWithConfig(formats, nil)
}

// NewHTTPClientWithConfig creates a new c e d a r core API HTTP client,
// using a customizable transport config.
func NewHTTPClientWithConfig(formats strfmt.Registry, cfg *TransportConfig) *CEDARCoreAPI {
	// ensure nullable parameters have default
	if cfg == nil {
		cfg = DefaultTransportConfig()
	}

	// create transport and client
	transport := httptransport.New(cfg.Host, cfg.BasePath, cfg.Schemes)
	return New(transport, formats)
}

// New creates a new c e d a r core API client
func New(transport runtime.ClientTransport, formats strfmt.Registry) *CEDARCoreAPI {
	// ensure nullable parameters have default
	if formats == nil {
		formats = strfmt.Default
	}

	cli := new(CEDARCoreAPI)
	cli.Transport = transport
	cli.AuthorityToOperate = authority_to_operate.New(transport, formats)
	cli.Budget = budget.New(transport, formats)
	cli.Component = component.New(transport, formats)
	cli.Contract = contract.New(transport, formats)
	cli.CostType = cost_type.New(transport, formats)
	cli.DataCenter = data_center.New(transport, formats)
	cli.Deployment = deployment.New(transport, formats)
	cli.DomainModel = domain_model.New(transport, formats)
	cli.Enumeration = enumeration.New(transport, formats)
	cli.Exchange = exchange.New(transport, formats)
	cli.Person = person.New(transport, formats)
	cli.Role = role.New(transport, formats)
	cli.Stakeholder = stakeholder.New(transport, formats)
	cli.SupportContact = support_contact.New(transport, formats)
	cli.System = system.New(transport, formats)
	cli.User = user.New(transport, formats)
	return cli
}

// DefaultTransportConfig creates a TransportConfig with the
// default settings taken from the meta section of the spec file.
func DefaultTransportConfig() *TransportConfig {
	return &TransportConfig{
		Host:     DefaultHost,
		BasePath: DefaultBasePath,
		Schemes:  DefaultSchemes,
	}
}

// TransportConfig contains the transport related info,
// found in the meta section of the spec file.
type TransportConfig struct {
	Host     string
	BasePath string
	Schemes  []string
}

// WithHost overrides the default host,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithHost(host string) *TransportConfig {
	cfg.Host = host
	return cfg
}

// WithBasePath overrides the default basePath,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithBasePath(basePath string) *TransportConfig {
	cfg.BasePath = basePath
	return cfg
}

// WithSchemes overrides the default schemes,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithSchemes(schemes []string) *TransportConfig {
	cfg.Schemes = schemes
	return cfg
}

// CEDARCoreAPI is a client for c e d a r core API
type CEDARCoreAPI struct {
	AuthorityToOperate authority_to_operate.ClientService

	Budget budget.ClientService

	Component component.ClientService

	Contract contract.ClientService

	CostType cost_type.ClientService

	DataCenter data_center.ClientService

	Deployment deployment.ClientService

	DomainModel domain_model.ClientService

	Enumeration enumeration.ClientService

	Exchange exchange.ClientService

	Person person.ClientService

	Role role.ClientService

	Stakeholder stakeholder.ClientService

	SupportContact support_contact.ClientService

	System system.ClientService

	User user.ClientService

	Transport runtime.ClientTransport
}

// SetTransport changes the transport on the client and all its subresources
func (c *CEDARCoreAPI) SetTransport(transport runtime.ClientTransport) {
	c.Transport = transport
	c.AuthorityToOperate.SetTransport(transport)
	c.Budget.SetTransport(transport)
	c.Component.SetTransport(transport)
	c.Contract.SetTransport(transport)
	c.CostType.SetTransport(transport)
	c.DataCenter.SetTransport(transport)
	c.Deployment.SetTransport(transport)
	c.DomainModel.SetTransport(transport)
	c.Enumeration.SetTransport(transport)
	c.Exchange.SetTransport(transport)
	c.Person.SetTransport(transport)
	c.Role.SetTransport(transport)
	c.Stakeholder.SetTransport(transport)
	c.SupportContact.SetTransport(transport)
	c.System.SetTransport(transport)
	c.User.SetTransport(transport)
}
