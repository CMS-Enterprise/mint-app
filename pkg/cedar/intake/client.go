// Package intake serves as an Anti-Corruption Layer (in Domain Driven Design
// parlance), keeping the implementation details of working with the
// downstream CEDAR Intake API well separated, so the rest of the internal
// workings of the EASi code are not using anything defined by the
// external API
package intake

import (
	"context"
	"fmt"
	"net/http"
	"strconv"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	"go.uber.org/zap"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/health_check"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/intake"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/translation"
	"github.com/cmsgov/easi-app/pkg/flags"
	"github.com/cmsgov/easi-app/pkg/models"
)

const (
	emitFlagKey = "emit-to-cedar"
	emitDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Intake API
func NewClient(cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		// this is the conditional way of stopping us from submitting to CEDAR; see EASI-1025
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(emitFlagKey, lduser, emitDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", emitFlagKey),
				zap.Bool("flagDefault", emitDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	hc := http.DefaultClient

	return &Client{
		emitToCedar: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc: hc,
	}
}

// Client represents a connection to the CEDAR Intake API
type Client struct {
	emitToCedar func(context.Context) bool
	auth        runtime.ClientAuthInfoWriter
	sdk         *apiclient.CEDARIntake
	hc          *http.Client
}

// CheckConnection hits the CEDAR Intake API `/healthcheck` endpoint to verify
// that our connection is functional
func (c *Client) CheckConnection(ctx context.Context) error {
	if !c.emitToCedar(ctx) {
		return nil
	}

	params := health_check.NewHealthCheckParamsWithContext(ctx)
	params.HTTPClient = c.hc

	resp, err := c.sdk.HealthCheck.HealthCheck(params, c.auth)
	if err != nil {
		return err
	}
	if resp.Payload == nil {
		return fmt.Errorf("no response body received")
	}
	return nil
}

// PublishSystemIntake sends a system intake to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishSystemIntake(ctx context.Context, si models.SystemIntake) error {
	intakeObject := translation.TranslatableSystemIntake(si)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishBusinessCase sends a business case to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishBusinessCase(ctx context.Context, bc models.BusinessCase) error {
	intakeObject := translation.TranslatableBusinessCase(bc)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishGRTFeedback sends an item of GRT feedback to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishGRTFeedback(ctx context.Context, feedback models.GRTFeedback) error {
	intakeObject := translation.TranslatableFeedback(feedback)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishAction sends an action to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishAction(ctx context.Context, action models.Action) error {
	intakeObject := translation.TranslatableAction(action)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishNote sends a note to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishNote(ctx context.Context, note models.Note) error {
	intakeObject := translation.TranslatableNote(note)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// private method for publishing anything that satisfies the translation.IntakeObject interface to CEDAR through the Intake API
func (c *Client) publishIntakeObject(ctx context.Context, model translation.IntakeObject) error {
	// constant values for now; may become non-constant if/when we revisit handling CEDAR validation errors

	// from the Swagger (cedar_intake.json, definitions/IntakeInput/properties/version):
	// "The version associated with the object in the body. This value can be incremented in the event a transaction needs to be resubmitted."
	const objectVersion = 1

	// from the Swagger (cedar_intake.json, paths/"/intake"/post/parameters/validatePayload):
	// "Determines if schema validation of the payload is performed syncronously before persisting the record or asyncronously after the record has been persisted"
	const isValidatedSynchronously = false // false for V1 schemas; this may change when we implement V2 schemas for EASI-1614

	id := model.ObjectID()
	objectType := model.ObjectType()

	if !c.emitToCedar(ctx) {
		appcontext.ZLogger(ctx).Info("snapshot publishing disabled", zap.String("object ID", id), zap.String("object type", objectType))
		return nil
	}

	input, err := model.CreateIntakeModel()

	if err != nil {
		appcontext.ZLogger(ctx).Error(
			"problem building cedar payload",
			zap.String("object ID", id),
			zap.String("object type", objectType),
			zap.Error(err),
		)
		return nil
	}

	params := intake.NewIntakeAddParamsWithContext(ctx)
	params.HTTPClient = c.hc
	params.Body = input

	params.ValidatePayload = strconv.FormatBool(isValidatedSynchronously)

	objectVersionStr := strconv.FormatInt(objectVersion, 10)
	params.Body.Version = &objectVersionStr

	resp, err := c.sdk.Intake.IntakeAdd(params, c.auth)
	if err != nil {
		return err
	}

	// TODO: we may need to read through the results for any "duplicate entry" error codes
	// that CEDAR _may_ send us, especially if we are re-transmitting entities... We
	// don't really care if there are dupes on their side, because we do NOT want to hold
	// any state on our side about who/what/when may have transmitted a previous copy
	// of the info to CEDAR.
	// appcontext.ZLogger(ctx).Info(fmt.Sprintf("results: %v", resp.Payload))
	_ = resp

	return nil
}
