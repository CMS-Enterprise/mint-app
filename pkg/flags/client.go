package flags

import (
	"context"
	"crypto/sha256"
	"fmt"
	"time"

	"github.com/cmsgov/mint-app/pkg/appcontext"

	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfiledata"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfilewatch"

	"github.com/cmsgov/mint-app/pkg/appconfig"
)

// DowngradeAssessmentTeamKey is the Flag Key in LaunchDarkly for downgrading assessment team users
const DowngradeAssessmentTeamKey = "downgradeAssessmentTeam"

// Config has all the parts for creating a new LD Client
type Config struct {
	Source         appconfig.FlagSourceOption
	Key            string
	Timeout        time.Duration
	FlagValuesFile string
}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	if config.Source == appconfig.FlagSourceLaunchDarkly {
		return ld.MakeClient(config.Key, config.Timeout)
	}

	defaultTimeout := 5 * time.Second

	if config.Source == appconfig.FlagSourceFile {

		ldConfig := ld.Config{
			DataSource: ldfiledata.DataSource().FilePaths(config.FlagValuesFile).Reloader(ldfilewatch.WatchFiles),
		}
		return ld.MakeCustomClient("fake_key", ldConfig, defaultTimeout)
	}

	// we default to an OFFLINE client for non-deployed environments
	return ld.MakeCustomClient("fake_offline_key", ld.Config{Offline: true}, 5*time.Second)
}

// Principal builds the LaunchDarkly user object for the
// currently authenticated principal.
func Principal(ctx context.Context) lduser.User {
	p := appcontext.Principal(ctx)
	key := UserKeyForID(p.ID())

	// this is a bit of a loose inference, assuming a user w/o Job Codes
	// is an Anonymous user. Over time, may want to consider adding
	// a `func Anonymous() bool` accessor to the authorizaion.Principal interface
	// definition instead of doing this inference
	authed := (p.AllowUSER() || p.AllowASSESSMENT())

	return lduser.
		NewUserBuilder(key).
		Anonymous(!authed).
		Build()
}

// UserKeyForID generates a user key from an ID
// we should not be using bare EUA IDs as identifiers to
// LaunchDarkly (per Jimil/ISSO), so we use a cryptographically
// secure one-way hash of the EUA ID as "key" for the LD User object.
func UserKeyForID(id string) string {
	h := sha256.New()
	_, _ = h.Write([]byte(id))
	return fmt.Sprintf("%x", h.Sum(nil))
}
