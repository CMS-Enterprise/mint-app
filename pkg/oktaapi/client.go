package oktaapi

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/okta/okta-sdk-golang/v2/okta"
	"github.com/okta/okta-sdk-golang/v2/okta/query"

	"github.com/cmsgov/mint-app/pkg/models"
)

// Client is a wrapper around github.com/okta/okta-sdk-golang/v2/okta Client type.
// The purpose of this package is to act as a drop-in replacement for the CEDAR LDAP API, so this package should implement the same
// methods that the Client interface defines in pkg/cedar/cedarldap/translated_client.go
type Client struct {
	oktaClient *okta.Client
}

// NewClient creates a Client
func NewClient(url string, token string) (*Client, error) {
	// TODO Do we need the "Context" response from okta.NewClient??
	_, oktaClient, oktaClientErr := okta.NewClient(context.TODO(), okta.WithOrgUrl(url), okta.WithToken(token))
	if oktaClientErr != nil {
		return nil, oktaClientErr
	}
	return &Client{
		oktaClient: oktaClient,
	}, nil
}

type oktaUserProfile struct {
	FirstName   string `json:"firstName"`
	LastName    string `json:"lastName"`
	DisplayName string `json:"displayName"`
	Email       string `json:"email"`
	Login       string `json:"login"`
}

func parseOktaProfileResponse(profile *okta.UserProfile) (*models.UserInfo, error) {
	// Create an okaUserProfile to return
	parsedProfile := &oktaUserProfile{}

	// Marshal the profile into a string so we can later unmarshal it into a struct
	responseString, err := json.Marshal(profile)
	if err != nil {
		return nil, err
	}

	// Unmarshal the string into the oktaUserProfile type
	json.Unmarshal(responseString, parsedProfile)

	returnInfo := &models.UserInfo{
		CommonName: parsedProfile.DisplayName,
		Email:      models.EmailAddress(parsedProfile.Email),
		EuaUserID:  parsedProfile.Login,
		FirstName:  parsedProfile.FirstName,
		LastName:   parsedProfile.LastName,
	}

	return returnInfo, nil
}

func (o *Client) FetchUserInfo(ctx context.Context, username string) (*models.UserInfo, error) {
	user, _, err := o.oktaClient.User.GetUser(ctx, username)
	if err != nil {
		return nil, err
	}

	profile, err := parseOktaProfileResponse(user.Profile)
	if err != nil {
		return nil, err
	}

	return profile, nil
}

func (o *Client) SearchCommonNameContains(ctx context.Context, searchTerm string) ([]*models.UserInfo, error) {
	filterString := fmt.Sprintf(`profile.firstName eq "%v"`, searchTerm)
	filter := query.NewQueryParams(query.WithFilter(filterString))

	filteredUsers, _, err := o.oktaClient.User.ListUsers(ctx, filter)
	if err != nil {
		fmt.Printf("Error Getting Users: %v\n", err)
	}

	users := make([]*models.UserInfo, len(filteredUsers))
	for idx, user := range filteredUsers {
		profile, err := parseOktaProfileResponse(user.Profile)
		if err != nil {
			return nil, err
		}
		users[idx] = profile
	}
	return users, nil
}
