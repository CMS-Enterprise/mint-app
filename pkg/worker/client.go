package worker

import (
	"fmt"

	faktory "github.com/contribsys/faktory/client"
)

// NewClient returns a new faktory Client.
func NewClient() (*faktory.Client, error) {
	cl, err := faktory.Open()
	if err != nil {
		return nil, fmt.Errorf("error opening faktory client: %w", err)
	}
	return cl, nil
}
