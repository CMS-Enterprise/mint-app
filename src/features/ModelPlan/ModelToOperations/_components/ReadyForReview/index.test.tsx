import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import { MtoStatus } from 'gql/generated/graphql';
import i18next from 'i18next';

import MTOReadyForReview from '.';

describe('MTO Ready for review', () => {
  const mockCloseModal = vi.fn();

  // ReactModel is throwing warning - App element is not defined. Please use `Modal.setAppElement(el)`.  The app is being set within the modal but RTL is not picking up on it
  // eslint-disable-next-line
  console.error = vi.fn();

  it('rendered correct in progress text', () => {
    const { getByText } = render(
      <MockedProvider>
        <MemoryRouter
          initialEntries={[
            '/models/0272ca43-1ec1-45a6-a06f-8e2def7f6888/collaboration-area/model-to-operations/matrix?view=milestones'
          ]}
        >
          <Route path="/models/:modelID/collaboration-area/model-to-operations/matrix">
            <MTOReadyForReview
              isOpen
              closeModal={mockCloseModal}
              status={MtoStatus.IN_PROGRESS}
            />
          </Route>
        </MemoryRouter>
      </MockedProvider>
    );
    expect(
      getByText(
        i18next.t('modelToOperationsMisc:readyForReview.headingInReview')
      )
    ).toBeInTheDocument();
    expect(
      getByText(
        i18next.t('modelToOperationsMisc:readyForReview.descriptionReady')
      )
    ).toBeInTheDocument();
    expect(
      getByText(i18next.t('modelToOperationsMisc:readyForReview.markAsReady'))
    ).toBeInTheDocument();
  });

  it('rendered correct ready for review text', () => {
    const { getByText } = render(
      <MockedProvider>
        <MemoryRouter
          initialEntries={[
            '/models/0272ca43-1ec1-45a6-a06f-8e2def7f6888/collaboration-area/model-to-operations/matrix?view=milestones'
          ]}
        >
          <Route path="/models/:modelID/collaboration-area/model-to-operations/matrix">
            <MTOReadyForReview
              isOpen
              closeModal={mockCloseModal}
              status={MtoStatus.READY_FOR_REVIEW}
            />
          </Route>
        </MemoryRouter>
      </MockedProvider>
    );
    expect(
      getByText(
        i18next.t('modelToOperationsMisc:readyForReview.headingInProgress')
      )
    ).toBeInTheDocument();
    expect(
      getByText(
        i18next.t('modelToOperationsMisc:readyForReview.descriptionInProgress')
      )
    ).toBeInTheDocument();
    expect(
      getByText(
        i18next.t('modelToOperationsMisc:readyForReview.markAsInProgress')
      )
    ).toBeInTheDocument();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <MockedProvider>
        <MemoryRouter
          initialEntries={[
            '/models/00000000-0000-0000-0000-000000000005/collaboration-area/model-to-operations/matrix?view=milestones'
          ]}
        >
          <Route path="/models/:modelID/collaboration-area/model-to-operations/matrix">
            <MTOReadyForReview
              isOpen
              closeModal={mockCloseModal}
              status={MtoStatus.IN_PROGRESS}
            />
          </Route>
        </MemoryRouter>
      </MockedProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
