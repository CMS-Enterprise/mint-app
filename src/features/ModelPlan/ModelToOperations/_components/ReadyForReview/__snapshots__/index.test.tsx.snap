// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MTO Ready for review > matches the snapshot 1`] = `<DocumentFragment />`;
